src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs

```diff
@@ -295,6 +313,88 @@ void ScheduleTask(DateTimeOffset utcNow)
             }
         }
 
+        internal readonly List<WeakReference<Stats>?> _allStats = new();
+        internal readonly Stats _accumulatedStats;
+
+        private (int, int) Sum()
+        {
+            int hits = 0, misses = 0;
+            lock (_allStats)
+            {
+                hits += _accumulatedStats.Hits;
+                misses += _accumulatedStats.Misses;
+                foreach (WeakReference<Stats>? wr in _allStats)
+                {
+                    if (wr is not null && wr.TryGetTarget(out Stats? stats))
+                    {
+                        hits += stats.Hits;
+                        misses += stats.Misses;
+                    }
+                }
+            }
+            return (hits, misses);
+        }
+
+        private ThreadLocal<Stats>? _stats;
+
+        private void Hit()
+        {
+            GetStats().Hits++;
+        }
+
+        private void Miss()
+        {
+            GetStats().Misses++;
+        }
+
+        private Stats GetStats()
+        {
+            if (_stats == null)
+            {
+                return Initialize();
+            }
+            return _stats!.Value!;
+
+            Stats Initialize()
+            {
+                var s = _stats = new ThreadLocal<Stats>(() => new Stats(this));
+                lock (_allStats)
+                {
+                    _allStats.Add(new WeakReference<Stats>(s.Value!));
+                }
+                return s.Value!;
+            }
+        }
+
+        internal sealed class Stats
+        {
+            public int Hits;
+            public int Misses;
+            private MemoryCache _memoryCache;
+            public Stats(MemoryCache memoryCache)
+            {
+                _memoryCache = memoryCache;
+            }
+            ~Stats()
+            {
+                var allStats = _memoryCache._allStats;
+                lock (allStats)
+                {
+                    for (int i = 0; i < allStats.Count; i++)
+                    {
+                        if (allStats[i] is WeakReference<Stats> wr && wr.TryGetTarget(out Stats? stats) && stats == this)
+                        {
+                            allStats[i] = null;
+                            break;
+                        }
+                    }
+                    allStats.RemoveAll(s => s is null);
+                    Interlocked.Add(ref _memoryCache._accumulatedStats.Hits, Hits);
```

Comment:
Nit: you don't need the interlocked operations because this code is already synchronized by the lock.